### PR TITLE
Fix app version update and upload

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:19805)](https://staging.teamyapp.com/teams/107/tasks/19805)

### CODE REVIEW

File: core/dao/sqldb/app_version.go [hunk 1]
```diff
@@ -68,6 +69,7 @@ func (a *AppVersion) FindAppVersionsByAppIDWithTx(ct context.Context, tx *transa
 			&appVersion.Status,
 			&appVersion.Locked,
 			&appVersion.IconURL,
```
Comment (naturetime10) [inline core/dao/sqldb/app_version.go]: Let's remove this field from DB and calculate file path at frontend.

File: core/dao/sqldb/app_version.go [hunk 2]
```diff
@@ -157,6 +163,8 @@ func (a *AppVersion) CreateAppVersion(ct context.Context, tx *transaction.Transa
 		appVersion.CreatedByUserID,
 		appVersion.Status,
 		appVersion.Locked,
+		appVersion.IconURL,
```
Comment (naturetime10) [inline core/dao/sqldb/app_version.go]:
```suggestion
```

File: core/dao/sqldb/app_version.go [hunk 3]
```diff
@@ -145,7 +149,9 @@ func (a *AppVersion) CreateAppVersion(ct context.Context, tx *transaction.Transa
 			$7,
 			$8,
 			$9,
-			$10
+			$10,
+			$11,
+			$12
```
Comment (naturetime10) [inline core/dao/sqldb/app_version.go]:
```suggestion
			$10,
			$11
```

File: core/dao/sqldb/app_version.go [hunk 4]
```diff
@@ -172,22 +180,24 @@ func (*AppVersion) UpdateAppVersion(ct context.Context, tx *transaction.Transact
...
-			created_at = $7
-			updated_at = $8
-		WHERE app_id = $7 AND number = $8;`,
+			status = $4,
+			locked = $5,
+			icon_url = $6,
+			created_at = $7,
+			updated_at = $8,
+			error_message = $9
+		WHERE app_id = $10 AND number = $11;`,
```
Comment (naturetime10) [inline core/dao/sqldb/app_version.go]:
```suggestion
			created_at = $6,
			updated_at = $7,
			error_message = $8
		WHERE app_id = $9 AND number = $10;`,
```

File: core/dao/sqldb/app_version.go [hunk 5]
```diff
@@ -172,22 +180,24 @@ func (*AppVersion) UpdateAppVersion(ct context.Context, tx *transaction.Transact
...
 		appVersion.AppName,
 		appVersion.HasUiExtension,
 		appVersion.Description,
 		appVersion.Status,
 		appVersion.Locked,
 		appVersion.IconURL,
```
Comment (naturetime10) [inline core/dao/sqldb/app_version.go]:
```suggestion
```

File: core/entity/app_version.go [hunk 6]
```diff
@@ -25,11 +26,12 @@ type AppVersion struct {
 	Status          AppVersionStatus
 	Locked          bool
 	IconURL         *string
```
Comment (naturetime10) [inline core/entity/app_version.go]:
```suggestion
```

File: core/entity/currency.go [hunk 7]
```diff
@@ -5,3 +5,7 @@ type Currency string
 const (
 	CurrencyUSD Currency = "USD"
 )
+
+var CurrencyStrToCurrency = map[string]Currency{
```
Comment (naturetime10) [inline core/entity/currency.go]:
```suggestion
var StringToCurrency = map[string]Currency{
```

File: core/mutation/create_app_version_changes.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
+		return err
+	}
+
+	c.notifiersPrepared = true
+	return nil
+}
+
+func (c *CreateAppVersionChange) Undo() *errs.Error {
+	return nil
```
Comment (naturetime10) [inline core/mutation/create_app_version_changes.go:56]: Should this be `nil`?

File: core/mutation/update_app_version.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		return err
+	}
+
+	u.notifiersPrepared = true
+	return nil
+}
+
+func (u *UpdateAppVersion) Undo() *errs.Error {
+	return nil
```
Comment (naturetime10) [inline core/mutation/update_app_version.go:56]: Should this be "nil"?

File: core/service/app.go [hunk 10]
```diff
@@ -877,7 +887,39 @@ func (a App) CreateAppPackageFileUploadSession(ct context.Context, appID uint64,
...
+		if err != nil {
+			return err
+		}
+
+		appVersion, err := a.appVersionDao.FindAppVersionByAppIDAndVersionNumberWithTx(ct, tx, appID, versionNumber)
+		if err != nil {
+			return err
+		}
+
+		if appVersion.Status != entity.AppVersionStatusInit && appVersion.Status != entity.AppVersionStatusError {
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
		if appVersion.Status != entity.AppVersionStatusInit && 
		appVersion.Status != entity.AppVersionStatusError {
```

File: core/service/app.go [hunk 11]
```diff
@@ -940,17 +982,79 @@ func (a App) FinishAppPackageFileUploadSession(ct context.Context, appID uint64,
...
+					return err
+				}
+
+				rtTx.AppendMutation(updateAppVersionMutation)
+				return nil
+			})
+
+			if transactionErr != nil {
+				a.logger.ErrorWithContext(ct, transactionErr)
+			}
```
Comment (naturetime10) [inline core/service/app.go:1057]:
```suggestion
			}
			
```

File: core/service/app.go [hunk 12]
```diff
@@ -1210,17 +1306,48 @@ func (a App) processManifestFile(ct context.Context, userID uint64, appID uint64
...
+			a.stateSyncer,
+			a.appVersionDao,
+			a.appDao,
+			appVersion,
+		)
+
+		err = updateAppVersionMutation.Execute(ct, tx)
+		if err != nil {
+			return err
+		}
```
Comment (naturetime10) [inline core/service/app.go:1352]:
```suggestion
		}
		
```

File: migrations/20240217213537_add-error-messsage-column.sql [hunk 13]
```diff
@@ -0,0 +1,8 @@
+-- +migrate Up
+ALTER TABLE app_version 
+ADD COLUMN error_message VARCHAR(100);
```
Comment (naturetime10) [inline migrations/20240217213537_add-error-messsage-column.sql]:
```suggestion
    ADD COLUMN error_message VARCHAR(100);
```

File: migrations/20240217213537_add-error-messsage-column.sql [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+
+-- +migrate Down
+ALTER TABLE app_version
+DROP COLUMN error_message;
+
```
Comment (naturetime10) [inline migrations/20240217213537_add-error-messsage-column.sql]:
```suggestion
```

File: migrations/20240217213537_add-error-messsage-column.sql [hunk 15]
```diff
@@ -0,0 +1,8 @@
+-- +migrate Up
+ALTER TABLE app_version 
+ADD COLUMN error_message VARCHAR(100);
+
+-- +migrate Down
+ALTER TABLE app_version
+DROP COLUMN error_message;
```
Comment (naturetime10) [inline migrations/20240217213537_add-error-messsage-column.sql]:
```suggestion
    DROP COLUMN error_message;
```

File: core/service/app.go [hunk 16]
```diff
@@ -940,17 +982,79 @@ func (a App) FinishAppPackageFileUploadSession(ct context.Context, appID uint64,
...
+				a.transactionFactory,
+				a.stateSyncer,
+				ct,
+			)
+
+			transactionErr := txCtx.WithTransactions(false, func(tx *cloudTransaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+				appVersion, err := a.appVersionDao.FindAppVersionByAppIDAndVersionNumberWithTx(ct, tx, appID, versionNumber)
+				if err != nil {
+					return err
+				}
```
Comment (naturetime10) [inline core/service/app.go:1032]:
```suggestion
				}
				
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/mutation/update_app_version.go (hunk 9 above)
Comment (duanxianbo) [reply core/mutation/update_app_version.go:56]: I think we would want to update with previous values, but we just did not implement this yet, all mutations undo is not implemented

File: core/mutation/create_app_version_changes.go (hunk 8 above)
Comment (duanxianbo) [reply core/mutation/create_app_version_changes.go:56]: same as above, but would be easier for creation, just delete the created entry, but we do not have this for any other create mutations

File: core/service/app.go [hunk 17]
```diff
@@ -1279,7 +1282,7 @@ func (a App) processManifestFile(ct context.Context, userID uint64, appID uint64
 		}
 
 		for currencyStr, price := range manifestData.Prices {
-			currency, ok := entity.CurrencyStrToCurrency[currencyStr]
+			currency, ok := entity.StringToCurrency[currencyStr]
```
Comment (naturetime10) [inline core/service/app.go:1285]:
```suggestion
			currency := entity.Currency(currencyStr)
```

File: core/entity/currency.go [hunk 18]
```diff
@@ -6,6 +6,6 @@ const (
 	CurrencyUSD Currency = "USD"
 )
 
-var CurrencyStrToCurrency = map[string]Currency{
+var StringToCurrency = map[string]Currency{
 	"USD": CurrencyUSD,
 }
```
Comment (naturetime10) [inline core/entity/currency.go]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/app.go (hunk 17 above)
Comment (duanxianbo) [reply core/service/app.go:1285]: what if user enter an currency that does not exist

File: core/api/gql/app_version_change.go [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+	"context"
+
+	"github.com/graph-gophers/graphql-go"
+	"github.com/teamyapp/cloud/libs/errs"
+	"github.com/teamyapp/teamy-backend/core/entity"
+)
+
+type AppVersionChange struct {
+	appVersionChange entity.AppVersionChange
+	deps             *Dependencies
```
Comment (naturetime10) [inline core/api/gql/app_version_change.go]:
```suggestion
	deps             *Dependencies
	appVersionChange entity.AppVersionChange
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**